### PR TITLE
Fixed #26645 -- Fixed potential makemessages failures on Windows.

### PR DESCRIPTION
**Pull request by ramiro**

### CODE REVIEW

File: django/core/management/commands/makemessages.py [hunk 1]
```diff
@@ -153,16 +153,45 @@ def cleanup(self):
...
+    """
+    Takes either a list of lines without EOL sequences (possibly already
+    returned by str.splitlines()) or a block of raw text that will be passed
+    through str.splitlines() to get Universal newlines treatment (has
+    precedence).
+
+    Returns the resulting block of text with normalised `\n` EOL sequences
+    ready to be written to disk using current platform native EOLs.
+    """
+    assert bool(lines_list is not None) != bool(raw_contents is not None)
```
Comment (ramiro) [inline django/core/management/commands/makemessages.py]: This is too clever. It tries to make sure only one of the two arguments is present, a logical xor operation.

Two possible alternatives:
* Simply don't check the condition, rely on the user reeading the docstring above.
* Replace it with a more readable check, e.g.:
```
assert lines_list is None or raw_contents is None
assert lines_list is not None or raw_contents is not None
```

File: django/core/management/commands/makemessages.py [hunk 2]
```diff
@@ -153,16 +153,45 @@ def cleanup(self):
...
+    if raw_contents is not None:
+        lines_list = raw_contents.splitlines()
+    # Ensure last line has its EOL
```
Comment (ramiro) [inline django/core/management/commands/makemessages.py]: We should perform this check about not trimming a traling EOL only when we've called `splitlines()` ourselves, so this entire block should be indented and additional level.

**Comment by ramiro** [comment]
Only commit is 7dc19ee is actually about fixing problems on Windows.

OTOH comits 7b8dae8 and c802a6c solve problems present on every platform so perhaps they should be on their own branches/PRs?

File: django/core/management/commands/makemessages.py [hunk 3]
```diff
@@ -153,16 +153,45 @@ def cleanup(self):
                 os.unlink(self.work_path)
 
 
+def build_text_file_content(lines_list=None, raw_contents=None):
+    """
+    Takes either a list of lines without EOL sequences (possibly already
```
Comment (timgraham) [inline django/core/management/commands/makemessages.py]: use pep257 verb style for new docstring "Take... Return..."

File: django/core/management/commands/makemessages.py [hunk 4]
```diff
@@ -153,16 +153,45 @@ def cleanup(self):
...
+    returned by str.splitlines()) or a block of raw text that will be passed
+    through str.splitlines() to get Universal newlines treatment (has
```
Comment (timgraham) [inline django/core/management/commands/makemessages.py]: universal (no capital)

File: django/core/management/commands/makemessages.py [hunk 5]
```diff
@@ -153,16 +153,45 @@ def cleanup(self):
...
+    precedence).
+
+    Returns the resulting block of text with normalised `\n` EOL sequences
```
Comment (timgraham) [inline django/core/management/commands/makemessages.py]: normalized

File: django/core/management/commands/makemessages.py [hunk 6]
```diff
@@ -153,16 +153,45 @@ def cleanup(self):
...
+    ready to be written to disk using current platform native EOLs.
```
Comment (timgraham) [inline django/core/management/commands/makemessages.py]: platform's

File: django/core/management/commands/makemessages.py (hunk 1 above)
Comment (timgraham) [inline django/core/management/commands/makemessages.py]: Maybe two separate functions would be easier to comprehend?

**Comment by timgraham** [comment]
I still see these failures:
Python 3.5:
```
======================================================================
FAIL: test_msgfmt_error_including_non_ascii (i18n.test_compilation.CompilationErrorHandling)
----------------------------------------------------------------------
django.core.management.base.CommandError: Execution of msgfmt failed: C:\Users\tim\code\django\tests\i18n\comma                               nds\locale\ko\LC_MESSAGES\django.po:24: 'msgstr' is not a valid Python brace format string, unlike 'msgid'. Rea                               son: In the directive number 0, '▒' cannot start a field name.
msgfmt: found 1 fatal error


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\tim\code\django\tests\i18n\test_compilation.py", line 187, in test_msgfmt_error_including_non_                               ascii
    call_command('compilemessages', locale=['ko'], verbosity=0)
  File "C:\Users\tim\AppData\Local\Programs\Python\Python35\Lib\contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "c:\users\tim\code\django\django\test\testcases.py", line 604, in _assert_raises_message_cm
    self.assertIn(expected_message, str(cm.exception))
AssertionError: "'\ufffd' cannot start a field name" not found in "Execution of msgfmt failed: C:\\Users\\tim\\                               code\\django\\tests\\i18n\\commands\\locale\\ko\\LC_MESSAGES\\django.po:24: 'msgstr' is not a valid Python brac                               e format string, unlike 'msgid'. Reason: In the directive number 0, '▒' cannot start a field name.\r\nmsgfmt: f                               ound 1 fatal error\r\n"

======================================================================
FAIL: test_special_char_extracted (i18n.test_extraction.BasicExtractorTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\tim\code\django\tests\i18n\test_extraction.py", line 219, in test_special_char_extracted
  File "C:\Users\tim\code\django\tests\i18n\test_extraction.py", line 87, in assertMsgId
  File "C:\Users\tim\code\django\tests\i18n\test_extraction.py", line 84, in _assertPoKeyword
AssertionError: None is not true : Could not find 'msgid "Non-breaking space▒:"' in generated PO file
```
Python 2.7:
```
======================================================================
ERROR: test_msgfmt_error_including_non_ascii (i18n.test_compilation.CompilationErrorHandling)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\tim\code\django\tests\i18n\test_compilation.py", line 179, in test_msgfmt_error_including_non_ascii
    call_command('compilemessages', locale=['ko'], verbosity=0)
  File "c:\users\tim\code\django\django\core\management\__init__.py", line 130, in call_command
    return command.execute(*args, **defaults)
  File "c:\users\tim\code\django\django\core\management\base.py", line 356, in execute
    output = self.handle(*args, **options)
  File "c:\users\tim\code\django\django\core\management\commands\compilemessages.py", line 103, in handle
    self.compile_messages(locations)
  File "c:\users\tim\code\django\django\core\management\commands\compilemessages.py", line 128, in compile_messages
    output, errors, status = popen_wrapper(args)
  File "c:\users\tim\code\django\django\core\management\utils.py", line 21, in popen_wrapper
    p = Popen(args, shell=False, stdout=PIPE, stderr=PIPE, close_fds=os.name != 'nt')
  File "C:\Users\tim\code\django\tests\i18n\test_compilation.py", line 175, in <lambda>
    with mock.patch('django.core.management.utils.Popen', lambda *args, **kwargs: Popen(*args, env=env, **kwargs)):
  File "C:\Python27\Lib\subprocess.py", line 710, in __init__
    errread, errwrite)
  File "C:\Python27\Lib\subprocess.py", line 958, in _execute_child
    startupinfo)
TypeError: environment can only contain strings

----------------------------------------------------------------------
```
I don't have a strong opinion about what to do with the unrelated commits. Creating a ticket for them never hurts. I guess they don't need to be backported to 1.10.x to get the build passing, unlike the other one.

**Comment by timgraham** [comment]
By "Unbreak" in the commit message, I'm not sure if you mean it as a synonym for "Fixed" or something else?

File: django/core/management/commands/makemessages.py (hunk 1 above)
Comment (ramiro) [inline django/core/management/commands/makemessages.py]: You've made me realize it doesn't need to handle both cases. We can handle one caling `\n.join(...)`.

I've simplified and renamed the function.

**Comment by ramiro** [comment]
> I still see these failures:
> Python 3.5:

Yes, they are unrelated to this issue. Hopefully I will be able to tackle them next.

**Comment by ramiro** [comment]
> I don't have a strong opinion about what to do with the unrelated commits. Creating a ticket for them never hurts. I guess they don't need to be backported to 1.10.x to get the build passing, unlike the other one.

I've opened #26670 for the related to the plural forms header one. Will do the same for the one about avoiding repeated entried in the locale paths list.

**Comment by timgraham** [comment]
LGTM

**Pull request closed without merging**